packages/block-library/src/image/edit.native.js

```diff
@@ -1,75 +1,130 @@
 /**
  * External dependencies
  */
-import { View, Image, TextInput } from 'react-native';
+import React from 'react';
+import { View, Image, TextInput, NativeEventEmitter } from 'react-native';
 import RNReactNativeGutenbergBridge from 'react-native-gutenberg-bridge';
 
+const gutenbergBridgeEvents = new NativeEventEmitter( RNReactNativeGutenbergBridge );
+
 /**
  * Internal dependencies
  */
 import { MediaPlaceholder, RichText, BlockControls } from '@wordpress/editor';
-import { Toolbar, ToolbarButton } from '@wordpress/components';
+import { Toolbar, ToolbarButton, Spinner } from '@wordpress/components';
 import { __ } from '@wordpress/i18n';
 
-export default function ImageEdit( props ) {
-	const { attributes, isSelected, setAttributes } = props;
-	const { url, caption } = attributes;
-
-	const onUploadPress = () => {
-		// This method should present an image picker from
-		// the device.
-		//TODO: Implement upload image method.
-	};
-
-	const onMediaLibraryPress = () => {
-		RNReactNativeGutenbergBridge.onMediaLibraryPress( ( mediaUrl ) => {
-			if ( mediaUrl ) {
-				setAttributes( { url: mediaUrl } );
-			}
-		} );
-	};
-
-	if ( ! url ) {
+const mediaUploadStateUploading = 1;
+const mediaUploadStateSucceeded = 2;
+const mediaUploadStateFailed = 3;
+
+export default class ImageEdit extends React.Component {
+	constructor( props ) {
+		super( props );
+
+		this.state = {
+			progress: 0,
+		};
+
+		this.mediaUpload = this.mediaUpload.bind( this );
+		this.addMediaUploadListener = this.addMediaUploadListener.bind( this );
+		this.removeMediaUploadListener = this.removeMediaUploadListener.bind( this );
+		this.finishMediaUploading = this.finishMediaUploading.bind( this );
+	}
+
+	mediaUpload( payload ) {
+		if ( payload.state === mediaUploadStateUploading ) {
+			this.setState( { progress: payload.progress } );
+		} else if ( payload.state === mediaUploadStateSucceeded || payload.state === mediaUploadStateFailed ) {
+			this.finishMediaUploading( payload );
+		}
+	}
+
+	finishMediaUploading( payload ) {
+		const { setAttributes } = this.props;
+
+		setAttributes( { url: payload.mediaUrl, id: payload.mediaId } );
+		this.setState( { progress: payload.progress } );
+		this.removeMediaUploadListener( payload.mediaId );
+	}
+
+	addMediaUploadListener( mediaId ) {
+		gutenbergBridgeEvents.addListener( 'mediaUpload' + mediaId, this.mediaUpload );
+	}
+
+	removeMediaUploadListener( mediaId ) {
+		gutenbergBridgeEvents.removeListener( 'mediaUpload' + mediaId, this.mediaUpload );
```

Comment:
I think it would be good to hide the implementation of adding/removing listeners inside the `react-native-gutenberg-bridge` module. Something similar to what we already have in [here](https://github.com/wordpress-mobile/gutenberg-mobile/blob/60256db1342732ce006cbaae0d5fb9c0793fed54/react-native-gutenberg-bridge/index.js#L24).

If we can also hide the `RNReactNativeGutenbergBridge.onMediaLibraryPress` it would be great too. Even though we didn't do it before, probably is good to have all the available communication methods in one place.

Reply:
Hey @etoledom, thanks for the comments. I have changed the code as we discussed.